### ISSUE TITLE
Voice system improvements

### ISSUE THREAD

**Issue by oaojmiguel** [multitheftauto/mtasa-blue#327]
**Is your feature request related to a problem? Please describe.**
- [ ] An excellent new feature for RP servers is the ability to check if a player has a mic plugged in (to show a custom icon or not). In my case, I'll display a mic icon (or not) besides the player's name.
- [x] Another excellent feature for RP servers is the ability to change the voice volume for each player based in a distance given. In doing this, we can implement the proximity voice chat system and some other cool stuffs.
- [x] Fix voice crashing and bugging out

**Describe the solution you'd like**
Enable both features or create them.

Mantis: https://bugs.mtasa.com/view.php?id=7168

**Comment by qaisjp** [comment]
Sounds reasonable. Please can you split this into two separate issues? Thanks!

**Comment by Einheit-101** [comment]
I already created 3D voice but it was complete garbage because of how silent MTA voice chat is (we might just delete voice entirely)

**Comment by Addlibs** [comment]
I’m pretty sure the second point is already implemented; setSoundVolume takes a player or sound element (this extends to most, if not all, sound-related functions) so you can interpolate the volume based on 3D distance.

And indeed, the voice chat is very quiet for some reason.

**Comment by PlatinMTA** [comment]
I don't know if it's only for me, but activating the voice chat sometimes crashes the client. I used the default voice resource so I don't think that was the problem, but i don't have any more information or idea why this happends.

And yes, the voice chat is really quiet.

**Comment by fusiion** [comment]
@PlatinMTA that also happened to me and all of my friends, activating the voice chat sometimes crashes the client, so it's not only you that's having this problem.

**Comment by patrikjuvonen** [comment]
Related Mantis issues: https://bugs.mtasa.com/view.php?id=7168

**Comment by botder** [comment]
Voice freezing the client has been fixed in 39c1ba00ae122f0393cf5f185033b17c10b392da

**Comment by PlatinMTA** [comment]
It would be nice if someone could work a little update to fix the voice sound issue.

**Comment by Dutchman101** [comment]
> It would be nice if someone could work a little update to fix the voice sound issue.

I think it's better to wait with individual improvement points until @Jusonex is done working on PR #895 which can solve multiple issues at once and offer new possibilities.

Besides that, possibly due to a combination of 39c1ba0 and various BASS library updates (or other unknown changes) over the past year, apart from the freezes we haven't observed a voice-related crash in crash stats for a long time. Although that may be a result of servers barely using it due to its reputation as unreliable and bad.

**Comment by PlatinMTA** [comment]
> I think it's better to wait with individual improvement points until @Jusonex is done working on PR #895 which can solve multiple issues at once and offer new possibilities.
> 
> Besides that, possibly due to a combination of [39c1ba0](https://github.com/multitheftauto/mtasa-blue/commit/39c1ba00ae122f0393cf5f185033b17c10b392da) and various BASS library updates (or other unknown changes) over the past year, apart from the freezes we haven't observed a voice-related crash in crash stats for a long time. Although that may be a result of servers barely using it due to its reputation as unreliable and bad.

I can assure you the crashes stopped when the update rolled up (at least none of my users crashed again for no apparent reason). I was asking cuz I'm working on a server that uses the voice system (TTT), but because it is really, really quiet and, as far as I know, you can't turn up the volume via resource some of my users can't really hear anything the other players are saying, so most of them use discord as an alternative, but this creates some problems.

I didn't search the code about the issue, but I guess it isn't that hard to just turn up a lil bit the volume, we already have an voice slider on the audio configuration anyway. Maybe I'm wrong, I don't really know.

At the state the voice is right now, yes, it can be used, it does not crash but some players can't hear anything, and that kinda sucks.

**Comment by Dutchman101** [comment]
> I was asking cuz I'm working on a server that uses the voice system (TTT), but because it is really, really quiet and, as far as I know, you can't turn up the volume via resource some of my users can't really hear anything the other players are saying, so most of them use discord as an alternative, but this creates some problems. 

> I didn't search the code about the issue, but I guess it isn't that hard to just turn up a lil bit the volume, we already have an voice slider on the audio configuration anyway. Maybe I'm wrong, I don't really know.

I see now.. if we can't expect the voice overhaul anytime soon, I would agree it's worth looking into the sound volume issue for who uses it in the meanwhile.

**Comment by qaisjp** [comment]
Principally I would say it should wait on #895, but practically I don't think it's going to happen anytime soon, and it's not worth waiting for it. Also, these features don't seem too difficult to implement.

- checking if a player has a mic plugged in can be achieved in a few different ways, and it's important to get the API for this right.

    It would be good if someone could do some research on how an API for this is achieved in other games or SDKs

- I don't think this is valid anymore, because `setSoundVolume` takes a player argument:

    > Another excellent feature for RP servers is the ability to change the voice volume for each player based in a distance given. In doing this, we can implement the proximity voice chat system and some other cool stuffs.

    Is the author satisfied with this outcome? It's not clear what else you're looking for here.

- "quiet voice" - this might be because human hearing in real life is not linear, it's logarithmic.

    [BASS_ChannelSetAttribute / BASS_ATTRIB_VOL](https://www.un4seen.com/doc/#bass/BASS_ATTRIB_VOL.html) has support for a `BASS_SLIDE_LOG` option, which is interesting!

    I can't find `BASS_SLIDE_LOG` anywhere in our code, so I suspect we're doing audio scaling very wrong right now.

**Comment by Einheit-101** [comment]
I implemented proximity voice long ago and removed it a week later, you cant even hear something when the player is close to you and it gets worse when they move a few meters.

**Comment by PlatinMTA** [comment]
setSoundVolume kinda fixes the issue, however i needed to put like 6.0 so players can really listen to each other well.

**Comment by Allerek** [comment]
https://github.com/multitheftauto/mtasa-blue/issues/1146
